tdms/tests/unit/field_tests/test_Field.cpp

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file test_Field.cpp
+ * @brief Unit tests for the Field class and its subclasses (ElectricField, MagneticField).
+ *
+ * NOTE: Interpolation methods are checked in a separate file for readability.
+ */
+#include "field.h"
+
+#include <complex>
+
+#include <catch2/catch_test_macros.hpp>
+
+#include "globals.h"
+#include "unit_test_utils.h"
+
+using namespace std;
+using namespace tdms_math_constants;
+
+const double tol = 1e-16;
+
+TEST_CASE("Field: normalise_volume") {
+
+  const int I_tot = 2, J_tot = 2, K_tot = 2;
+  // create a field (note electric or magnetic should be sufficient here as we are testing base class method)
+  ElectricField F(I_tot, J_tot, K_tot);
+  F.allocate_and_zero();
+
+  // set the angular norm to be 1 + i
+  F.angular_norm = 1. + IMAGINARY_UNIT * 1.;
+  // set components
+  for (int i = 0; i < I_tot; i++) {
+    for (int j = 0; j < J_tot; j++) {
+      for (int k = 0; k < K_tot; k++) {
+        // set x component to be 1 + i, so should normalise to 1.
+        F.real.x[k][j][i] = 1.;
+        F.imag.x[k][j][i] = 1.;
+        // set y component to be 2, so should normalise to be 1 - i
+        F.real.y[k][j][i] = 2.;
+        F.imag.y[k][j][i] = 0.;
+        // keep z components as 0
+      }
+    }
+  }
+
+  // now normalise
+  F.normalise_volume();
+  // check normalisation was correct
+  bool x_normalised_correctly = true;
+  bool y_normalised_correctly = true;
+  bool z_normalised_correctly = true;
+  for (int i = 0; i < I_tot; i++) {
+    for (int j = 0; j < J_tot; j++) {
+      for (int k = 0; k < K_tot; k++) {
+        // set x component to be 1 + i, so should normalise to 1.
+        x_normalised_correctly =
+                (x_normalised_correctly &&
+                 (abs(F.real.x[k][j][i] - 1. + IMAGINARY_UNIT * (F.imag.x[k][j][i] - 0.)) < tol));
+        // set y component to be 2, so should normalise to be 1 - i
+        y_normalised_correctly =
+                (y_normalised_correctly &&
+                 (abs(F.real.y[k][j][i] - 1. + IMAGINARY_UNIT * (F.imag.y[k][j][i] + 1.)) < tol));
+        // keep z components as 0
+        z_normalised_correctly =
+                (z_normalised_correctly &&
+                 (abs(F.real.z[k][j][i] - 0. + IMAGINARY_UNIT * (F.imag.z[k][j][i] - 0.)) < tol));
```

Comment:
```suggestion
        x_normalised_correctly &= (abs(F.real.x[k][j][i] - 1. + IMAGINARY_UNIT * (F.imag.x[k][j][i] - 0.)) < TOLERANCE);
        // set y component to be 2, so should normalise to be 1 - i
        y_normalised_correctly &= (abs(F.real.y[k][j][i] - 1. + IMAGINARY_UNIT * (F.imag.y[k][j][i] + 1.)) < TOLERANCE);
        // keep z components as 0
        z_normalised_correctly &= (abs(F.real.z[k][j][i] - 0. + IMAGINARY_UNIT * (F.imag.z[k][j][i] - 0.)) < TOLERANCE);
```

Reply:
But also, you could do the `REQUIRE` for each iteration in the loop? Or is that slower?

Reply:
A *bit* slower. It was more so that we don't end up bloating how many checks we are doing in this test. 

My ethos was I'm testing whether each of `x`, `y`, `z` have had _all_ their components normalised correctly, so I should have 3 `REQUIRES`. Doing the `REQUIRES` in the loop just bloats the number of checks we do; I feel $2/3$ tests failing is more informative than `X / I_tot*J_tot*K_tot`.